### PR TITLE
Modify standard error messages of pydantic

### PR DESCRIPTION
**Pull request by phackstock**
Proposed modification of the standard error message for ValidationErrors from pydantic to include the file name(s) where errors originated. 
The original error message is as follows:
```python
ValidationError: 1 validation error for RegionAggregationMapping
__root__
  Conflict between (renamed) native regions and aggregation mapping to common regions: ['common_region_1'] in tests/data/region_aggregation/illegal_mapping_conflict_regions.yaml (type=value_error)
```
The goal was to get this:
```python
ValidationError: 1 validation error for RegionAggregationMapping in `region_aggregation/illegal_mapping_conflict_regions.yaml`:
__root__
  Conflict between (renamed) native regions and aggregation mapping to common regions: ['common_region_1' (type=value_error)
```

In order to achieve this behavior we must first understand the structure of pydantic error handling.
As far as I understand it, pydantic collects all errors raised during validation and then raises them together in a single error wrapping class called [`ValidationError`](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/error_wrappers.py#L50). The individual errors raised in the validators should **not** be ValidationErrors but instead `ValueError`, `TypeError` or `AssertionError` (see [documentation](https://pydantic-docs.helpmanual.io/usage/models/#error-handling) for details). 
This is what makes changing the error message tricky because in this two tier error structure we can usually only influence the layer of Value- Type- or AssertionErrors and not of the wrapper ValidationError.
Looking into the source of `ValidationError`, the behavior of [`ValidationError.__str__`](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/error_wrappers.py#L70) would need to be modified. 
Luckily, altering the behavior of a certain object in python even if the object in question is imported from a third party library (see [this stackoverflow post](https://stackoverflow.com/questions/19545982/monkey-patching-a-class-in-another-module-in-python) for details). This is called monkey-patching. As `__str__` is a special ('dunder') method we can only monkey patch it on the class and not on the individual instance level (see [this stackoverflow post](https://stackoverflow.com/questions/60062100/is-it-possible-to-override-a-class-call-method)).
However, monkey patching a function as fundamental as `__str__` is potentially dangerous as it could break compatibility if changes are made upstream. Therefore, I opted to deep copy the original string method and only add functionality while still mostly using the original.
The second piece of the puzzle is getting the information about the file from the lower level of the ValueError to the higher wrapper class ValidationError. In order to make this easier I created a custom error ([pydantic documentation](https://pydantic-docs.helpmanual.io/usage/models/#custom-errors)). To keep things as simple as possible I've only created one custom error so far `DoubleNativeRegionError(PydanticValueError)`. As this error is invoked it is handed a file parameter which is then used in the modified `ValidationError.__str__` to produce the desired output.

At the end of `region_mapping_models.py` I have put an example that illustrates the newly formatted error.

Would love to hear your thoughts @danielhuppmann.
Also @meksor, would be cool to hear what you think about it. Is this complete overkill or closer to the minimum necessary to achieve the desired behavior?

### CODE REVIEW

**Comment by meksor** [comment]
How does your error class behave if there is errors in multiple files? Does it just use the file the first error was in? 
I guess the most elegant solution to this would be to get the behavior you want merged into pydantic itself. Not sure if that isn't too much work though.

**Comment by phackstock** [comment]
Good point, right now I just use the first file. I think even for multiple errors though the file will always be the same since the `RegionAggregationMapping` class constitutes a 1 to 1 mapping to a yaml mapping file. So I should only ever have one file per ValidationError even if I have multiple different errors

File: nomenclature/region_mapping_models.py [hunk 1]
```diff
@@ -5,7 +5,37 @@
...
+        )
+    return original__str__(self)
+
+
+# Overwrite the original __str__ with new__str__
+pydantic.error_wrappers.ValidationError.__str__ = new__str__
+
+# Custom error class since we need to get the file information to ValidationError
+# See for details: https://pydantic-docs.helpmanual.io/usage/models/#custom-errors
+class DoubleNativeRegionError(PydanticValueError):
```
Comment (danielhuppmann) [inline nomenclature/region_mapping_models.py]: Suggest to rename to `DuplicateRegionError` or `RegionConflictError` - Double isn't quite intuitive...

**Comment by danielhuppmann** [comment]
Seems like a decent and pragmatic solution to me - unless @meksor has design objections, I'd say write a test (including testing the error message so that we are alerted when the upstream pedantic message changes) and merge.

**Comment by phackstock** [comment]
Alright, I'll do that. Good idea with testing the upstream message.

**Comment by phackstock** [comment]
Since this has overlaps with iamconsortium/nomenclature#34, I'll merge this now and we can continue the discussion over there.

**Pull request merged**